app/Middleware/Queue/SubmissionService.php

```diff
@@ -161,20 +171,66 @@ public function getConsumerName()
      * Handles the incoming message
      *
      * @param Message $message
-     * @return void
+     * @return bool
      * @throws \Exception
      */
-    public function doSubmit(Message $message)
+    private function doSubmit(Message $message)
     {
         try {
-            $fh = fopen($message->file, 'r');
-            $config = Config::getInstance();
-            $config->set('CDASH_REMOTE_ADDR', $message->ip);
-            do_submit($fh, $message->project, null, $message->md5, $message->checksum);
+            Config::getInstance()->set('CDASH_REMOTE_ADDR', $message->ip);
+
+            if ($this->localProcessing) {
+                // Local execution, process an open file.
+                $fh = fopen($message->file, 'r');
+            } else {
+                // Remote execution, pass the filename and CDash will retrieve
+                // its contents.
+                $fh = basename($message->file);
+            }
+            $handler = do_submit($fh, $message->project, null, $message->md5, $message->checksum);
+            if (is_object($handler) && property_exists($handler, 'backupFileName')) {
+                $this->backupFileName = $handler->backupFileName;
+            }
         } catch (\Exception $e) {
             Log::getInstance()->error($e);
             throw $e;
         }
+        return true;
+    }
+
+    /**
+     * Request that the web server delete or achive a submission file.
+     *
+     * @param Message $message
+     * @return void
+     */
+    public function delete(Message $message)
+    {
+        if ($this->localProcessing) {
+            return;
+        }
+
+        // When remotely executing we tell the server to delete the file
+        // after it has been parsed.
+        $query_args = ['filename' => basename($message->file)];
+        if ($this->backupFileName) {
+            // If the handler provided us with a descriptive backup filename
+            // we pass this on to the server in case it is configured to rename
+            // (rather than delete) processed submission files.
+            $query_args['dest'] = basename($this->backupFileName);
+        }
+        $url = Config::getInstance()->get('CDASH_BASE_URL') .
+            '/api/v1/deleteSubmissionFile.php';
+        $this->getHttpClient()->request('DELETE', $url,
```

Comment:
Now we can unit test this unit of work. Well almost. You'll need to have the ability to setHttpClient, e.g.:
```php
public function setHttpClient(GuzzleHttp\ClientInterface $client) 
...
```
In doing so your unit test can test can now mock ClientInterface (implemented by Guzzle\Http\Client), in your phpunit test e.g.:
```php
...
$mock_client = $this->getMockBuilder(GuzzleHttp\ClientInterface::class)
  ->getMockForAbstractClass(); // determines abstract/interface methods and creates stubs for them

$mock_client->expects($this->once())
  ->method('request')
  ->with('DELETE', $url, ['query' => $query_args]);

$sut = (new SubmissionService('doSubmit'))
  ->setHttpClient($mock_client);

$sut->delete($message);
```

I realize that in moving to Laravel we probably no longer have a need for this class so not sure that you have any interest in writing more unit tests for it. However, IMO any new code—especially new code that already has unit tests associated with it—should include full coverage of the new code.

Reply:
> However, IMO any new code—especially new code that already has unit tests associated with it—should include full coverage of the new code

Hard to argue with that!  I'll look into adding a couple new test cases for this PR.